### PR TITLE
[Tutorial] Quickstart - Running an Open Solver

### PR DESCRIPTION
**Pull request by bh2smith**
This guide walks the reader through all the components required to run your own local instance of the open-solver. In brief, that is (Installation, Configuration and Execution)

The last few sections contain advanced configuration instruction such as orderbook fiters.

### CODE REVIEW

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]:
```suggestion
git clone https://github.com/gnosis/dex-services.git
```

Otherwise:
```
$ git clone git@github.com:gnosis/dex-services.git
Cloning into 'dex-services'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: This repo is private, no user outside of Gnosis can clone it. Is this intended?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]:
```suggestion
sudo docker-compose build --build-arg SOLVER_BASE=gnosispm/dex-open-solver:master stablex-debug
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: I would move this explanation to where the counterintuitive behavior will become apparent.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: I don't understand this explanation. Is `SOLVER_BASE` use to match orders and the runtime-specified solver for something else?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: Maybe worth mentioning 1=mainnet, 4=Rinkeby?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: What is "my solver"? Runtime argument to what? The env variable is nice however, I'd put it first and maybe even ignore the argument.

Comment (fedgiac) [reply docs/tutorial-run-solver.md]: Also, I suppose:
```suggestion
Copy this orderbook file into `dex-services/target` and point your solver to this file via the runtime argument `--orderbook-file` or environment vairable `ORDERBOOK_FILE`. That is, you can append `ORDERBOOK_FILE=$(pwd)/dex-services/target/stablex_orderbook_mainnet.bin` to your `.env` file.
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: Which folder are we in?
I assume that every command is run from the folder from where you run `git clone`.
I don't know about MacOS, but on Linux you'd either user `$PWD` or `$(pwd)`, not `$(PWD)`.

So, maybe this? But I don't know Docker, and the path after the colon looks like a path inside Docker, so maybe the second `$(pwd)` is not necessary?
```suggestion
sudo docker-compose run -v $(pwd)/:$(pwd)/dex-services stablex-debug
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]:
```
error: 'open-solver' isn't a valid value for '--solver-type <solver-type>'
	[possible values: BestRingSolver, NaiveSolver, OpenSolver, StandardSolver]

	Did you mean 'OpenSolver'?
```
Since it's not recommended, I'd put the recommended way to do it first. It could also be a subsection of the next section, after changing the titles.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]:
```suggestion
cargo run --bin driver -- --solver-type OpenSolver --network-id $NETWORK_ID --node-url $ETHEREUM_NODE_URL --private-key $PRIVATE_KEY --orderbook-file /app/dex-services/target/stablex_orderbook_mainnet.bin
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]:
```suggestion
cargo run --bin driver -- --solver-type OpenSolver --network-id $NETWORK_ID --node-url $ETHEREUM_NODE_URL --private-key $PRIVATE_KEY
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: All env variables should be set inside the Docker image. From the tutorial I thought it was enough to set them in the base system. After this command it's good to source `.env`. (Which should be in the `dex-services` folder.)

Error I got otherwise:
```
2020-09-09T14:01:06.980Z ERRO [services_core::logging] thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: web3 error: Transport error: CouldntResolveHost: couldn't resolve host name
```

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: This should be done after moving into the docker instance, the folder of the current system is independent of the one of Docker.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: Note that a `.env` file outside the container in `dex-services` is read but not preserved in the container, so you need another `.env` in the container.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: I'm constantly being provided with alternatives where I don't really get why I have a choice instead of being told what to do. I'd stick to a single "standard" way to do thing, otherwise I'm losing the thread.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: Why not making it part of `.env`?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: I also got a lot of
```
2020-09-09T14:23:48.891Z WARN [services_core::price_estimation::clients::generic_client] failed to retrieve services_core::price_estimation::clients::dexag::api::DexagHttpApi prices for token ID 34 (aUSDT): failed to get price from dexag

Caused by:
    0: failed to parse JSON '"{\"error\":\"Error getting price\"}"'
    1: invalid type: string "{\"error\":\"Error getting price\"}", expected struct Price at line 1 column 37
2020-09-09T14:23:48.892Z WARN [services_core::price_estimation::clients::generic_client] failed to retrieve services_core::price_estimation::clients::dexag::api::DexagHttpApi prices for token ID 41 (aMKR): failed to get price from dexag
```
Is it because I just ran the first command you provided?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: I went through each step, this line is never needed.

**Review by fedgiac — COMMENTED** [review]
Nice, the tutorial is fairly simple considering how involved the whole process actually is.

I didn't run it for long enough to have a solution submitted however, so I don't know if the open solver is playing nice with `dex-services`.

Note on methodology: I fired up a fresh virtual machine and tried to follow all steps in order, trying to keep implicit steps to a minimum. So every command change I described was made because otherwise the command wouldn't run as expected.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: This is the open solver repo and it should be public unless I have the wrong link.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Sudo shouldn’t be necessary here and I would very much like to avoid having to use it at all costs.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: This is not the case for me when I mounted the project inside the container. That is, if the `.env` file exists within the project repo at the time we run 

```
docker-compose run -v $(PWD)/:/app/dex-services stablex-debug
```

it will also be copied into the container and be there when you need it.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Because I don’t want to provide my infura key.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Nah. These warning are because the price feed has no price.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: If I hadn't used sudo with docker-compose:
```
$ docker-compose run -v $(pwd)/:$(pwd)/dex-services stablex-debug
ERROR: Couldn't connect to Docker daemon at http+docker://localhost - is it running?

If it's at a non-standard location, specify the URL with the DOCKER_HOST environment variable.
```
This is a fresh, default installation of Docker and `docker-compose`.
If we assume the users are familiar with Docker, then sure, we can also assume they know what's going on with this kind of useless error message and then they either add `sudo` themselves like I did or set up their Docker to work rootles. I wouldn't rely on the latter however. To my understanding [rootless mode](https://docs.docker.com/engine/security/rootless/) is just a fig leaf to cover the need of Docker for superuser access, and has [no concrete security benefits over using sudo](https://docs.docker.com/engine/security/security/#docker-daemon-attack-surface).

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: I'd add a "warning warning", since these messages make it look like something might be going wrong.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]:
```suggestion
git clone git@github.com:gnosis/dex-open-solver.git
```

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: In any case, I would rather put a comment below saying something like "your docker-compose installation may require you to use sudo here". How do you like the sound of that?

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: solver base is the solver image that is built into the image. Each solver image might have its own "solver type" to choose from at run time. Would this explanation make things more clear?

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: I don't really see the benefit here to using `$PWD` as opposed to `/app` considering `/app` is hard coded as the `PWD` of the container anyway.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: We are in `dex-services`, perhaps I should make this more clear in the beginning. As for `$(PWD)`, this was taken as is specified in the `dex-services` README file and I would prefer to leave it as is for now, but I will try it out with the other suggestions. Also, again, `/app` is gonna stay.

Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Right, the env file should be sourced from within the container. I will try and make this more clear.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: I'm sorry, but this is not how I have been running things on my end. I will try and execute it this way as well and take your suggestion if it does indeed work this way.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Very strange. I will look into this.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: I would be in favour of sourcing an `.env` file so the run command is as short and sweet as possible. What do you think?

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Yes. I could also mention to use different log filter so these aren't even seen.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: Works for me!
(Is Docker rootless by default in Macs?)

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: I see, one specifies the Docker image, which can contain multiple solver( type)s.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: You are right, I was confused at first because you don't start the Docker image until later.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: Definitely keep `\app`!
About `$(PWD)`, maybe it's something Mac specific? If I run the command I get the error I'd expect: 
```
bash: PWD: command not found...
```
Note that `docker-compose` starts correctly anyway, so maybe the whole `$(PWD)` part can be simply left out. I'm not familiar enough with Docker that I can tell what the `-v` "volume" option is supposed to look like, I see that @fleupold  last edited that line, so maybe he knows?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: Strange, all the time I propose this change is because otherwise it fails with the error I described before. Maybe some recent change modified it?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: Ok, then this is most likely what `$(PWD)` is doing. What do you see if you run `echo $(PWD)`?

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: Definitely agree! I very much prefer `.env` file to very long command-line arguments.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [reply docs/tutorial-run-solver.md]: I see, they are all concrete values so `INFURA_KEY=<thing>` wouldn't be great.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: On macOS, Docker is installed in a place that I own. I guess that means rootless.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]:
```
➜  dex-docs git:(integration_tutorial) echo $(PWD)
/Users/admin/Projects/gnosis/dex-docs
➜  dex-docs git:(integration_tutorial) echo $PWD
/Users/admin/Projects/gnosis/dex-docs
```

I guess they are both the same. We could ask @fleupold if he had something particular in mind at the time this was included in the README.

Comment (fedgiac) [reply docs/tutorial-run-solver.md]: In Bash `$(command)` is substituted with the output of `command`, so that you can do something like:
```
$ echo "The current timestamp is $(date +%s)."
The current timestamp is 1599823115.
```
So the reason why `$(PWD)` is failing for me is that `PWD` is not an existing command. Assuming that the syntax doesn't change in your terminal, then it'd mean that there is a `PWD` command available in MacOS, in case you should be able to run it just by typing `PWD` in the console. (Alternatively, maybe commands are not case sensitive? `pwd` would be a valid command for me.)
By the way, `$PWD` (or equivalent but safer, `${PWD}`) works also for me. `$NAME` is substituted with the content of the `NAME` variable, and there happens to be a valid `PWD` variable set in Bash containing the path of the current directory.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Yes, this was a freshly introduced config change introduced just last week. 

https://github.com/gnosis/dex-services/pull/1413

I will make the necessary changes.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: I think for now I will just remove this since moving it will require substantial rewording. Will check to see if this is still necessary as I parse through the latter sections.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Yes this was changed last week. I have linked to the PR in a different resolved comment.

File: docs/tutorial-run-solver.md
Comment (bh2smith) [reply docs/tutorial-run-solver.md]: Actually, this is something I did before moving into the docker instance. It shouldn't really matter in which order this happens since the `/app` directory was mounted into the image from the project folder.

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: Nice, works perfectly!

File: docs/tutorial-run-solver.md
Comment (fedgiac) [inline docs/tutorial-run-solver.md]: Nice and short!

**Review by fedgiac — APPROVED** [review]
Looks good, nice work!

File: docs/tutorial-run-solver.md
Comment (nlordell) [reply docs/tutorial-run-solver.md]: `$(PWD)` is a ZSH thing, I think we should stick to Bash for compatibility.

Comment (fedgiac) [reply docs/tutorial-run-solver.md]: In case you are curious: I got hold of a Mac and was able to test this. Apparently `PWD` is indeed a command in MacOs, so `$()` is working in `zsh` as expected for Bash. Interestingly, `pwd` is a "shell builtin" while `PWD` is a standalone executable.
In my Linux machine, `PWD` doesn't exist and `pwd` is a standalone executable.

File: docs/tutorial-run-solver.md
Comment (nlordell) [inline docs/tutorial-run-solver.md]: it might be worth stating that this only works from within the docker container, and how to get there (unless you have the open solver installed on your host system).

File: docs/tutorial-run-solver.md
Comment (nlordell) [inline docs/tutorial-run-solver.md]: Maybe indicate that this is run from within the docker container.

**Review by nlordell — APPROVED** [review]
Looks good!

**Review by keikreutler — CHANGES_REQUESTED** [review]
I reviewed and pushed changes.

File: docs/tutorial-solver.md
Comment (fedgiac) [inline docs/tutorial-solver.md]: Just two typos I noticed:
```suggestion
and [here](https://rinkeby.etherscan.io/tx/0xef93563c9c79708a613fb77978bff974672679270f9b51f98c19a8ce90d35260) is an example of a successful solution submission on Rinkeby.
```

**Review by fedgiac — APPROVED** [review]

File: docs/tutorial-solver.md
Comment (bh2smith) [inline docs/tutorial-solver.md]: Usually, in markdown, each sentence is on its own line. This is so that any changes made to one sentence are easily seen in the changelog 
```suggestion
The following three environment variables are _required_ when running the dex-services project. 
Note that the account corresponding to the private key should be funded with sufficient ETH to pay gas for solution submission.
```

File: docs/tutorial-solver.md
Comment (bh2smith) [inline docs/tutorial-solver.md]:
```suggestion
Copy this orderbook file into `dex-services/target` and append `ORDERBOOK_FILE=/app/dex-services/target/stablex_orderbook_mainnet.bin` to your `.env` file from the previous section. 
Technically, the orderbook file can be saved anywhere, but we have chosen `target` here since it is flagged as an untracked directory in the project's `.gitignore`.
```

File: docs/tutorial-solver.md
Comment (bh2smith) [inline docs/tutorial-solver.md]:
```suggestion
or, equivalently, for the simplest and most robust experience, first source your configuration file, and then run the driver without any additional runtime arguments.
That is, from within the Docker container:
```

File: docs/tutorial-solver.md
Comment (bh2smith) [inline docs/tutorial-solver.md]: I'm not overly fond of this change. This is non-standard terminology
```suggestion
where `.env_rinkeby` contains the following, minimal, parameters:
```

**Comment by bh2smith** [comment]
Everything looks good here @keikreutler, just waiting on your approval and possibly merge. If possible, for future reviews, I would prefer if your changes were marked via the interface as suggestions so that I can see them and commit them.

**Pull request merged**
